vpr/src/route/rr_graph_uxsdcxx_serializer.h

```diff
@@ -627,13 +628,15 @@ class RrGraphSerializer final : public uxsd::RrGraphBase<RrGraphContextTypes> {
 
     inline void set_node_loc_side(uxsd::enum_loc_side side, int& inode) final {
         auto node = (*rr_nodes_)[inode];
+        auto& device_ctx = g_vpr_ctx.device();
```

Comment:
I suggest not to use any global variable in this file.
This data structures in this head file caches the routing resource graph. As you see, the ``rr_nodes_`` in the line above is the cached version.
You can add the RRGraphView to the data cache:
https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/c18b2f24cecd2bdfc57c1b0ad20828932cb07433/vpr/src/route/rr_graph_uxsdcxx_serializer.h#L250-L297
And then use it in the same as the line above.

When refactoring is finished, we remove the data cache for ``rr_nodes``, ``rr_node_indices``, etc.